### PR TITLE
AssemblyScript 17 upgrade

### PR DESCRIPTION
**Pull request by jtenner**
@willemneal Please review. I think we have some issues with the assembly package.

I also included a small fix to the snapshot parser.

### CODE REVIEW

File: packages/cli/src/run.ts [hunk 1]
```diff
@@ -95,7 +95,12 @@ export function run(cliOptions: Options, compilerArgs: string[]): void {
     }
 
     /** Next, collect the loader, and assert it has an instantiateSync method. */
-    let loader = require(path.join(assemblyScriptFolder, "lib", "loader"));
+    let loader;
+    try {
+      loader = require(path.join(assemblyScriptFolder, "lib", "loader"));
+    } catch (ex) {
+      loader = import(path.join(assemblyScriptFolder, "lib", "loader"));
```
Comment (jtenner) [inline packages/cli/src/run.ts:102]: Hey I thought `import()` returns a promise..?

**Review by jtenner — COMMENTED** [review]
Honestly, these changes all look good. I'll get around to checking these all myself soon.

File: packages/cli/src/run.ts (hunk 1 above)
Comment (willemneal) [reply packages/cli/src/run.ts:102]: Ah you're right. Then I don't know why this fixed it and why it type checked fine.

Comment (willemneal) [reply packages/cli/src/run.ts:102]: Really I should just to `require(..."/umd")`

File: packages/core/__tests__/__snapshots__/RTrace.spec.ts.snap [hunk 2]
```diff
@@ -234,14 +234,36 @@ exports[`RTrace output Overall Statistics: RTrace testserrors 1`] = `Array []`;
...
 exports[`RTrace output Overall Statistics: a group reference counting error rtraceStart 1`] = `7`;
 
-exports[`RTrace output Overall Statistics: a group reference counting errorerrors 1`] = `Array []`;
-
-exports[`RTrace output Overall Statistics: allocationCount 1`] = `60`;
-
-exports[`RTrace output Overall Statistics: errors 1`] = `Array []`;
-
-exports[`RTrace output Overall Statistics: freeCount 1`] = `50`;
+exports[`RTrace output Overall Statistics: a group reference counting errorerrors 1`] = `
```
Comment (jtenner) [inline packages/core/__tests__/__snapshots__/RTrace.spec.ts.snap:241]: We should take a look at this. What is this?

@willemneal

File: packages/core/src/test/TestContext.ts [hunk 3]
```diff
@@ -1499,4 +1501,8 @@ export class TestContext {
       this.reflectedValueCache[reflectedValueID].stringify(stringifyOptions),
     );
   }
+
+  onresize(_block: number, _size: number): void {
```
Comment (jtenner) [inline packages/core/src/test/TestContext.ts:1505]: The next pull request needs to implement the `RTrace` class. We bump the current version to 5.0, and then implement RTrace support in `5.1`.

File: packages/snapshots/src/parser/grammar.ne [hunk 4]
```diff
@@ -1,12 +1,14 @@
...
 
-start -> _ lines _ {% d => {
+start -> _ (lines _):? {% d => {
   const results = d[1];
   const map = new Map<string, string>();
-  for (let i = 0; i < results.length; i++) {
-    const [key, value] = results[i];
-    if (map.has(key)) throw new Error("Invalid snapshot.");
-    map.set(key, value);
+  if (results) {
```
Comment (jtenner) [inline packages/snapshots/src/parser/grammar.ne:6]: This bug fix helps load empty snapshots. A looming bug.

**Comment by jtenner** [comment]
closed in favor of #319

**Pull request closed without merging**